SPEC: Add loaded-via in loadTable API

singhpk234 opened the pull request:
### About the change 

This change proposes loaded-via in loadTable request, which is expected to contain the FQN of the view (only 2 part identifier), which a rest catalog would expect from the client based on that it knows the table is being loaded in the context of view (view referencing the table) so that table can an action accordingly.

This would be really helpful in following cases : 

Supporting the security for views, i.e definer / invoker mode : 
`Definer` mode means the access to table should be authorized against the principal which created the view, this will be a very normal case where one would want to create a view and grant access to view but not to the underlying table. 

`Invoker` mode means the access to table should be authorized against the principal which which is calling the loadTable essentially what happens in default. 

when a view is defined in a definer mode security, `loaded-via` would help give catalog proper signal that this loadTable is happening in the context of view (i.e view is referencing the table) so the catalog can reverse lookup who the creator was and what the security mode is defined, and take proper authZ action. 


### Reference Implementation 

TODO: trying spark 

The problem is SparkCatalog when doing loadTable just get the identifier, the view context which we got in the analysis phase need to made passed through all the way to icebergCatalog it hold, i have an option with SparkSession.getActive() in the spark Catalog and set in the session context to be used to making the call, still thinking of something better,

singhpk234 commented on open-api/rest-catalog-open-api.yaml:
```diff
@@ -960,6 +960,15 @@ paths:
+          description:
+            The fully qualified name of the view (i.e namespace name and view name) that referenced the table.
+            
+            nested namespaces are separated by a unit separator (`0x1F`) byte.
```
IMHO, i understand being explicit would be ideal but i think it would be a lot of terminology we would be including here, if we include this in the spec, for example 
> Authorization

we only talk of Authentication in IRC, privileges / grants which is what we will authorize against, requires settling its independent discussion like one here [[doc](https://docs.google.com/document/d/1KmIDbPuN6IYF0nWs9ostXIB9F4b8iH3zZO0hjgs1lm4/edit?tab=t.0#heading=h.jgpnl0ksivvu)],.

> as a Trusted Engine identity as determined by the catalog’s trust policy

never the less Trusted Engine as a concept should be defined in IRC first as well, before we reference here, like what do we mean by this, How does a catalog understands a compute is trusted, TBH i am unsure if we want to define it here in the first place, like can my OAuth token wrap this info, can mTLS tell me this, it need not be a `Policy` necessarily. 

Considering the above, I am a bit hesitant in including this gotcha in spec (i already included this in the proposal), would love you know your thoughts considering above. 

Nevertheless hear others in context of spec definition too.

My POV is this is a context we are passing from the client (just like any-other context like UA), hey this table is being loaded in the context of this view, without explicitly stating how a catalog should use it, it implies that trust needs be there if i were to take a AuthZ decision based on that.

rdblue replied:
I don't think that the spec should have requirements about how this is used by the service. This is informational and we should state that it should be correct. That is, we should say when a client should set this to and what information it provides to the service, but we don't place requirements about authorization in the REST spec so I don't see a reason to do it here.

flyrain commented on open-api/rest-catalog-open-api.yaml:
```diff
@@ -960,6 +960,15 @@ paths:
             type: string
             enum: [ all, refs ]
+        - in: query
+          name: referenced-by
```
Can we call it something like `referenced-by-view` since it can only be referenced by a view here? I understand that the description already explicitly says that it's a view, but people can still misuse it given a generic name like `referenced by`.

rdblue commented on open-api/rest-catalog-open-api.yaml:
```diff
@@ -960,6 +960,15 @@ paths:
           schema:
             type: string
             enum: [ all, refs ]
+        - in: query
```
This is informational so it is up to the service, right? I don't think that we need to require specific behavior.

rdblue commented on open-api/rest-catalog-open-api.yaml:
```diff
@@ -960,6 +960,15 @@ paths:
+          description:
+            The fully qualified name of the view (i.e namespace name and view name) that referenced the table.
+            
+            nested namespaces are separated by a unit separator (`0x1F`) byte.
```
I think we also need to specify how to join the namespace and the table name, right?

singhpk234 replied:
yes it does :), added it !

singhpk234 commented:
Pending on nested-namespace representation discussion to conclude : https://github.com/apache/iceberg/pull/14448/files

nastra commented on open-api/rest-catalog-open-api.yaml:
```diff
@@ -980,6 +980,32 @@ paths:
+            Multiple view identifiers are separated by commas. Servers should split the parameter value on comma characters
+            to parse individual view identifiers. If view names contain commas, they must be percent-encoded as %2C.
+
+            Example with multiple views - prod%1Fanalytics.monthly_view,prod%1Fanalytics.quarterly_view
```
nit: maybe you want to mention here that prod%1Fanalytics refers to a nested namespace

nastra commented on open-api/rest-catalog-open-api.yaml:
```diff
@@ -980,6 +980,32 @@ paths:
+
+            Each view identifier has the format namespace.viewName where multipart namespace parts must be separated
+            by the namespace separator as indicated via the /config override namespace-separator, which defaults to
+            the unit separator 0x1F byte (url encoded %1F). To be compatible with older clients, servers must use
```
```suggestion
            the unit separator 0x1F byte (url encoded `%1F`). To be compatible with older clients, servers must use
```

nastra commented on open-api/rest-catalog-open-api.yaml:
```diff
@@ -980,6 +980,32 @@ paths:
+            by the namespace separator as indicated via the /config override namespace-separator, which defaults to
+            the unit separator 0x1F byte (url encoded %1F). To be compatible with older clients, servers must use
+            both the advertised separator and 0x1F as valid separators when decoding namespaces.
+            The namespace-separator should be provided in a url encoded form.
```
```suggestion
            The `namespace-separator` should be provided in a url encoded form.
```

those comments are so that we're consistent across the OpenAPI spec on how we highlight fields

gaborkaszab commented on open-api/rest-catalog-open-api.yaml:
```diff
@@ -980,6 +980,32 @@ paths:
+        - in: query
+          name: referenced-by
+          description:
+            A comma-separated list of fully qualified view names (namespace and view name) representing the view
```
Probably this has been discussed so asking for my information: this param is a comma-separated string now. Have you considered introducing it as an array of TableIdentifiers or String?

gaborkaszab reviewed (commented):
I think this is fine, just left some nits or questions

qqqttt123 commented:
Thanks for your great work! I have som concern about the security risk because it relies on client-provided information without server-side validation. The server cannot verify if the referenced view actually exists or legitimately references the table being loaded, potentially allowing permission bypasses.

For example:
Tom creates a view `A` using SQL, has the privilege to read table `B` and table `C`.
```
CREATE VIEW A 
SQL SECURITY DEFINER
AS SELECT * FROM B WHERE department = 'Engineering';
```
Jerry has the privilege to read the view `A`, if he mocks a request load table C referenced by view `A`, it will produce a security risk.

We would better add more constraints in the `createVIew` request. We should add required table identifiers in the request. So we can validate the table identifiers in the server side.

singhpk234 commented on open-api/rest-catalog-open-api.yaml:
```diff
@@ -980,6 +980,30 @@ paths:
+            as defined in `/config`).
+
+            The namespace and view name are separated by a dot. When parsing, the dot separator is located by
+            finding the first dot character after the last namespace separator.
```
Thank you for raising this ! i can see how this can be confusing and unintentionally be an odd restriction.

> the last namespace part can't contain `.`

How about we say first `.` we reading the FQN of the view name when reading it from backwards, everything on the left is the namespace and right of the dot is viewName ? 

OR 

Just simply repurpose the namespace separator and just say everything to the right of the last namespace separator is view and left is the namespace 


I believe based on the email thread the feedback is to use namespace seperator, just wanted to double confirm, before I push the spec change.

rdblue commented on open-api/rest-catalog-open-api.yaml:
```diff
@@ -980,6 +980,30 @@ paths:
+          name: referenced-by
+          description:
+            A comma-separated list of fully qualified view names (namespace and view name) representing the view
+            reference chain when a table is loaded via a view. The list should be ordered with the outermost view
```
> Feels like it would be a security hole?

I'm not sure I follow the case where this could be a security hole. Any time you get the permissions of a DEFINER, you must have access to the DEFINER view. Wouldn't it be strange if the catalog's intent was to nest an INVOKER view inside a DEFINER view in order to protect data referenced by the INVOKER? And I don't think it's a hole if that's the case because the catalog is what gets to decide (at least with the referenced-by chain) what the behavior is.

RussellSpitzer replied:
Sorry, I mean more like a practice I would not encourage.

I feel like if you have an invoker, that is a signal that the data behind the invoker is very sensitive and only certain users should have it. If you add a definer on top of the invoker view it's like you are bypassing that. Better to build a definer view in the first place

singhpk234 commented:
pushed [d18f648](https://github.com/apache/iceberg/pull/13810/commits/d18f648cafcd23b54181385d70d85dc62462242e) did the following : 
1. Made referenced-by a reusable param 
2. Added the referenced-by in the loadView too, since in nested view a view may be referencing another view 
3. Added referenced-by in the loadCreds endpoint too, since this is where we would giving creds for READ to table and we need to be aware that the credential call is in context of the this, mostly for long running jobs. 

Apologies i was about to count votes and realized just now, would request you all please take another pass, the language is almost entirely same as what was approved !

c-thiel reviewed (approved):
Thanks for the addition and driving this @singhpk234, LGTM!
